### PR TITLE
10 Tips for Creating Robust Serverless Components

### PR DESCRIPTION
**Pull request by eahefnawy**

### CODE REVIEW

File: posts/2019-08-12-10-tips-creating-robust-serverless-components.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+  - guides-and-tutorials
+authors:
+  - EslamHefnawy
+---
+
+<iframe width="560" height="315" src="https://www.youtube.com/embed/ts26BVuX3j0" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+
+[Serverless Components](https://github.com/serverless/components) allow anyone to design and build specific use cases and share them with the world, without having to be tied to a single pattern or a single opinion. We then released 20+ components to bootstrap these uses cases, and we’re constantly building more. One of our core design goals is simplicity, so we produced an incredibly simple core and components API that makes creating components a breeze, and we’ve covered them all in a [simple one-page documentation](https://github.com/serverless/components/blob/master/README.md).
+
+In this article, we will give you comprehensive practice advice (not docs) on creating robust Serverless Components based on our experience building all these components over the past year, to help as you start creating components for your own unique use cases. So let’s dig in..
```
Comment (ac360) [inline posts/2019-08-12-10-tips-creating-robust-serverless-components.md]: Suggestion:   In this article, we will give you practical advice (not docs)...

File: posts/2019-08-12-10-tips-creating-robust-serverless-components.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+### 1. Start with the Outcome
+
+Serverless Components is an outcomes framework. Every single component produces a specific outcome. Those outcomes could be low-level (like an S3 bucket), or high level (like a complete cloud storage solution). Before you even start building your component, you need to decide on what that outcome is and how much abstraction you want to provide for the best UX.
```
Comment (ac360) [inline posts/2019-08-12-10-tips-creating-robust-serverless-components.md]: +1 "Serverless Components is an outcomes framework"

Suggestion:  Change (like a complete cloud storage solution) to (an image processing feature).

File: posts/2019-08-12-10-tips-creating-robust-serverless-components.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+Once you know what outcome you’re after, you’ll need to think about your component’s inputs and outputs, as well as what the defaults are. We recommend that you minimize the inputs at first, then you can iteratively add more config if at all needed, simply because removing inputs later is a breaking change, but adding inputs is not.
+
+If at all possible, aim for a zero config component that would use sane defaults to deploy even a demo of that outcome. One example for a zero-config component is the aws-s3 bucket. If you don’t specify any inputs to that component, it’ll just create an accelerated bucket for you with a randomly generated name.
+
+### 2. See Real-World Examples
+
+While the components core API is simple, the cloud provider you’re working with (ie. AWS) might not be super straightforward. Fortunately there are some repeated patterns that we have seen after creating 20+ serverless components. We recommend that you browse through the available component repos on Github and check which one is close to what you’re building. Just open up the repo and check the code and see how everything fits together.
```
Comment (ac360) [inline posts/2019-08-12-10-tips-creating-robust-serverless-components.md]: Suggestion:  add links to existing Components you recommend reviewing.

File: posts/2019-08-12-10-tips-creating-robust-serverless-components.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+All component code is open source, and we are striving to keep it clean, simple and easy to follow. After you look around you’ll be able to understand how our core API works, how we interact with external APIs, and how we are reusing other components. Which brings us to the next point…
+
+### 3. Reuse Existing Components
+
+Serverless Components were designed from the ground up to be sharable and reusable. You can compose components together to deliver higher level component with minimum code. For example, the website component uses the aws-s3 component to deploy and upload to the S3 bucket, and it also uses the domains component to setup the custom domain. Likewise, you can reuse the website component, add to it your own frontend code (ie. an entire React app) to create a configurable/reusable landing page. That’s exactly what the chat-app component does.
```
Comment (ac360) [inline posts/2019-08-12-10-tips-creating-robust-serverless-components.md]: Suggestion:  Add code samples.  This article is a big block of text so far.  Always seek to show, don't tell.  Don't tell them composition is easy, show them via a code sample.

Suggestion:  Add links when you reference other components.

File: posts/2019-08-12-10-tips-creating-robust-serverless-components.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+In many cases, especially if you’re creating a low level component, you’ll have to provision resources in some cloud provider (ie. AWS). In that case, you will likely need to choose a name or some sort of an identifier to your resource (ie. bucket name). If the resource name is an input, you will need to be aware that a resource of that name might already exist and handle that edge case. You will also need to handle the case of the user changing the name input, in which case you should delete the old resource with the old name (that was saved in state in a previous deployment), and create a new one with the new name. We recommend that you create the new resource before deleting the old one, in case something went wrong during creation.
+
+However if your use cases allows it, we recommend that you use random names for better UX and to avoid collisions completely, and then save that in the local state. We also recommend to use the contextId in the resource name as a form of tagging, as it groups all resources deployed from your component with that identifier. We have some helper functions for that, and that could look something like this:
+
+<code>
```
Comment (ac360) [inline posts/2019-08-12-10-tips-creating-robust-serverless-components.md]: Where is the code?

File: posts/2019-08-12-10-tips-creating-robust-serverless-components.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+Keep in mind however what we mentioned earlier about keeping most of the state in the provider. You can’t completely trust the local state, so you better verify the state of the resources on the provider, either by making a get request, or as mentioned in the previous section, by using a try/catch strategy of the removal (if it doesn’t exist, it’s already removed)
+
+### 9. Save Time by Using Core Utils
+
+When you first create your component, you’ll need to extend the Component class from the @serverless/core npm package. In that package, there are plenty of useful utilities that you can use to make your life easier. For example, we have a Utils.sleep() function that pauses runtime for a few seconds before moving on. This is extremely helpful when dealing with cloud infrastructure as it might take a while to create some resources before we make further operations on it. There are also file system, parsing and packaging utilities that you can use when dealing with files that you might need to upload to the cloud. Here’s a quick example of how to use some of those utilities
+
+<code>
```
Comment (ac360) [inline posts/2019-08-12-10-tips-creating-robust-serverless-components.md]: Code?

**Comment by eahefnawy** [comment]
Thanks @ac360 ... I had resolved and pushed some of these issues, but they weren't reflected here for some reason 🤔 ... I've made further updates and added more code snippets as you suggested.

**Pull request merged**